### PR TITLE
automatic publishing to Pypi and TestPypi using Github Actions

### PR DESCRIPTION
**Pull request by rgutzen**
Setting up automatic publishing of the Python package, following the [Python packaging guide](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/).
Each commit to master triggers an upload to TestPypi to ensure a working publishing pipeline. Each tagged commit to master triggers an upload to Pypi for an official release.

Additionally, this PR bumps the version number to 0.1.1

### CODE REVIEW

**Comment by rgutzen** [comment]
I tested the workflow by successfully publishing version 0.1.1 to TestPypi by pushing to the `feature/automatic-publishing` branch. So, now after changing back the workflow settings to only publish on pushes to master, I expect it to work analogously. I didn't explicitly test the publishing to Pypi with tagged commit, but as this uses the same mechanism and trusted publisher settings I'm confident that this also works as expected.

**Comment by rgutzen** [comment]
@mdenker, what's review of this new feature?

File: .github/workflows/publish-to-pypi-and-testpypi.yml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+
+
+  publish-to-pypi:
+    name: >-
+      Publish Python 🐍 distribution 📦 to PyPI
```
Comment (mdenker) [inline .github/workflows/publish-to-pypi-and-testpypi.yml:35]:
```suggestion
    name: Publish Python 🐍 distribution 📦 to PyPI
```

File: .github/workflows/publish-to-pypi-and-testpypi.yml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
+    needs:
+      - build
+    runs-on: ubuntu-latest
+    environment:
```
Comment (mdenker) [inline .github/workflows/publish-to-pypi-and-testpypi.yml:40]:
```suggestion

    environment:
```

File: .github/workflows/publish-to-pypi-and-testpypi.yml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      name: pypi
+      url: https://pypi.org/p/cobrawap
+    permissions:
```
Comment (mdenker) [inline .github/workflows/publish-to-pypi-and-testpypi.yml:43]:
```suggestion

    permissions:
```

File: .github/workflows/publish-to-pypi-and-testpypi.yml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      id-token: write  # IMPORTANT: mandatory for trusted publishing
+    steps:
```
Comment (mdenker) [inline .github/workflows/publish-to-pypi-and-testpypi.yml:45]:
```suggestion

    steps:
```

**Review by mdenker — APPROVED** [review]
I did not test this code, as I think I could not do more than what you had already described in the PR comments regarding testing the upload on the feature branch. However, I feel the code looks ok, however, there is one thing I am not so sure about.

If I understand correctly, this will create a Pypi release and a github release on each push with a tag.  So this means, we cannot tag anything but versions on master. Could this be dangerous in some situations, or should one check that the ref starts with `/refs/tags/v?.?.?` or so?

All in all, I think this looks great -- and at least for the github releases, no harm will be done in case there is still a bug.

Only minor cosmetic changes apart from this.

**Comment by rgutzen** [comment]
@mdenker, that's a good point. I added to the corresponding if statement that the tag must start with a "v" to trigger the publishing with pypi, which is far from foolproof but as much as I could do with yaml.
I think this might do for now, when we keep in mind not too start other tags with "v".
Other suggestions welcome!

**Comment by cosimolupo** [comment]
@rgutzen why not using regular expressions to parse the “v…” tags really intended to label versions?

**Comment by rgutzen** [comment]
@cosimolupo, regex can't be directly integrated into the workflow yaml syntax. Copilot suggests including an extra step to check the tag format like this:
```
    - name: Check tag format
      id: check_tag
      run: |
        if [[ "${{ github.ref }}" =~ ^refs/tags/v[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
          echo "::set-output name=TAG_MATCH::true"
        else
          echo "::set-output name=TAG_MATCH::false"
        fi
```
Though, tt needs to be double-checked though that it actually does what it claims.

**Comment by cosimolupo** [comment]
Well, this line
>         if [[ "${{ github.ref }}" =~ ^refs/tags/v[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
does actually use regex. This is exactly what I was suggesting. Of course then we need to test it.

**Comment by mdenker** [comment]
I think the regexing is a good suggestion -- is this super easy to implement? Then let's do it, otherwise, I'll merge this quick check for v* and then we include a more extensive check in the next PR.

**Pull request merged**